superset/db_engine_specs/superset.py

```diff
@@ -0,0 +1,343 @@
+
+
+def has_rowid(method: F) -> F:
+    @wraps(method)
+    def wrapper(self: "SupersetShillelaghAdapter", *args: Any, **kwargs: Any) -> Any:
+        # pylint: disable=protected-access
+        if not self._rowid:
+            raise ProgrammingError(
+                "Can only modify data in a table with a single, integer, primary key"
+            )
+        return method(self, *args, **kwargs)
+
+    return cast(F, wrapper)
+
+
+# pylint: disable=too-many-instance-attributes
+class SupersetShillelaghAdapter(Adapter):
+
+    """
+    A shillelagh adapter for Superset tables.
+
+    Shillelagh adapters are responsible for fetching data from a given resource,
+    allowing it to be represented as a virtual table in SQLite. This one works
+    as a proxy to Superset tables.
+    """
+
+    safe = True
+
+    type_map: Dict[Any, Type[Field]] = {
+        bool: Boolean,
+        float: Float,
+        int: Integer,
+        str: String,
+        datetime.date: Date,
+        datetime.datetime: DateTime,
+        datetime.time: Time,
+    }
+
+    @staticmethod
+    def supports(uri: str) -> bool:
+        # An URL for a table has the format superset.database[.catalog][.schema].table,
+        # eg, superset.examples.birth_names
+        parsed = urllib.parse.urlparse(uri)
+        parts = parsed.path.split(".")
+        return 3 <= len(parts) <= 5 and parts[0] == "superset"
+
+    @staticmethod
+    def parse_uri(uri: str) -> Tuple[str, Optional[str], Optional[str], str]:
+        parsed = urllib.parse.urlparse(uri)
+        parts = parsed.path.split(".")
+        if len(parts) == 3:
+            return parts[1], None, None, parts[2]
+        if len(parts) == 4:
+            return parts[1], None, parts[2], parts[3]
+        return tuple(parts[1:])  # type: ignore
+
+    def __init__(
+        self,
+        database: str,
+        catalog: Optional[str],
+        schema: Optional[str],
+        table: str,
+        username: str,
+    ):
+        self.database = database
+        self.catalog = catalog
+        self.schema = schema
+        self.table = table
+        self.username = username
+
+        self._rowid: Optional[str] = None
+        self._allow_dml: bool = False
+        self._set_columns()
+
+    @classmethod
+    def get_field(cls, python_type: Any) -> Field:
+        class_ = cls.type_map.get(python_type, Blob)
+        return class_(filters=[Equal, Range], order=Order.ANY, exact=True)
+
+    def _set_columns(self) -> None:
+        from superset.models.core import Database
+
+        database = (
+            db.session.query(Database).filter_by(database_name=self.database).first()
+        )
+        if database is None:
+            raise ProgrammingError(f"Database not found: {self.database}")
+        self._allow_dml = database.allow_dml
+
+        # verify permissions
+        # set user since the adapter runs in a different thread
+        g.user = security_manager.get_user_by_username(self.username)
```

Comment:
@dpgaspar I still need to set `g.user` here, since this is running in a separate thread. But at least we're setting the user to a username from the main thread, instead of using `flask_login.current_user`.

Have we considered changing `security_manager.raise_for_access` to take a username? That would simplify things here.

Reply:
I used the same approach from #19999 and passed username, which is then used by the security manager iff `g.user` is undefined.